vertical-pod-autoscaler/pkg/updater/eviction/pods_eviction_restriction.go

```diff
@@ -156,6 +159,61 @@ func (e *podsEvictionRestrictionImpl) Evict(podToEvict *apiv1.Pod, eventRecorder
 	return nil
 }
 
+// canDelete checks if the container is in a CrashLoopBackOff state and if it was OOMKilled.
+func (e *podsEvictionRestrictionImpl) canDelete(containerStatus apiv1.ContainerStatus) bool {
+	return containerStatus.LastTerminationState.Terminated != nil &&
+		containerStatus.LastTerminationState.Terminated.Reason == "OOMKilled" &&
+		containerStatus.State.Waiting != nil &&
+		containerStatus.State.Waiting.Reason == "CrashLoopBackOff"
+}
+
+// EvictViaDelete sends deletion instruction to api client. Returns error if pod cannot be deleted or if client returned error
+// Does not check if pod was actually deleted after termination grace period.
+func (e *podsEvictionRestrictionImpl) EvictViaDelete(podToEvict *apiv1.Pod, eventRecorder record.EventRecorder) error {
+	cr, present := e.podToReplicaCreatorMap[getPodID(podToEvict)]
+	if !present {
+		return fmt.Errorf("pod not suitable for eviction %v : not in replicated pods map", podToEvict.Name)
+	}
+
+	if !e.CanEvict(podToEvict) {
+		return fmt.Errorf("cannot evict pod %v : eviction budget exceeded", podToEvict.Name)
+	}
+
+	deleted := false
+	// Try to delete the Pod if it's CrashLooping because of OOM.
+	// The Pod is dead anyway so there is no reason to respect any potential PDBs.
+	for _, containerStatus := range podToEvict.Status.ContainerStatuses {
+		if e.canDelete(containerStatus) {
+
+			err := e.client.CoreV1().Pods(podToEvict.Namespace).Delete(context.TODO(), podToEvict.Name, metav1.DeleteOptions{})
+			if err != nil {
+				klog.Errorf("failed to delete pod %s/%s, after failed eviction, error: %v", podToEvict.Namespace, podToEvict.Name, err)
+				return err
+			}
+
+			deleted = true
+			eventRecorder.Event(podToEvict, apiv1.EventTypeNormal, "DeletedByVPA",
+				"Pod was deleted by VPA Updater to apply resource recommendation.")
+			break
```

Comment:
If I read this loop right it will attempt to delete 0 or 1 containers.

- 0 containers if canDelete doesn't return true for any container,
- 1 container otherwise (returns error if delete fails, breaks if it succeeds)

Is this the intended behavior (as opposed to attempting to delete all containers until we run into an error)?

If so I'd make it into its own function that returns `(bool, error)` to make this easier to read.

Reply:
It is the intended behaviour, because we cannot delete individual containers, but only the whole pod. So as soon as the pod is deleted, there is no reason to continue for the rest of the containers in the pod.